tests/tcp.rs

```diff
@@ -23,7 +23,99 @@ fn assert_error_kind<T>(res: io::Result<T>, kind: io::ErrorKind) {
 }
 
 async fn bind() -> std::result::Result<TcpListener, std::io::Error> {
-    TcpListener::bind((IpAddr::from(Ipv4Addr::UNSPECIFIED), PORT)).await
+    TcpListener::bind((Ipv4Addr::UNSPECIFIED, PORT)).await
+}
+
+#[test]
+fn connects_within_a_host() -> Result {
+    let mut sim = Builder::new().build();
+
+    sim.client("server", async {
+        let listener = bind().await?;
+        tokio::spawn(async move {
+            loop {
+                let _ = listener.accept().await;
+            }
+        });
+
+        TcpStream::connect(("server", PORT)).await?;
+        Ok(())
+    });
+
+    sim.run()
+}
+
+#[test]
+fn connects_within_a_localhost() -> Result {
+    let mut sim = Builder::new().build();
+
+    sim.client("server", async {
+        let listener = TcpListener::bind((Ipv4Addr::LOCALHOST, PORT)).await?;
+        tokio::spawn(async move {
+            loop {
+                let _ = listener.accept().await;
+            }
+        });
+
+        TcpStream::connect((Ipv4Addr::LOCALHOST, PORT)).await?;
+
+        assert_error_kind(
+            TcpStream::connect((Ipv6Addr::LOCALHOST, PORT)).await,
+            io::ErrorKind::ConnectionRefused,
+        );
+
+        Ok(())
+    });
+
+    sim.client("serveripv6", async {
+        let listener = TcpListener::bind((Ipv6Addr::LOCALHOST, PORT)).await?;
+
+        tokio::spawn(async move {
+            loop {
+                let _ = listener.accept().await;
+            }
+        });
+
+        TcpStream::connect((Ipv6Addr::LOCALHOST, PORT)).await?;
+
+        assert_error_kind(
+            TcpStream::connect((Ipv4Addr::LOCALHOST, PORT)).await,
+            io::ErrorKind::ConnectionRefused,
+        );
+
+        Ok(())
+    });
+
+    sim.run()
+}
+
+#[test]
+fn doesnt_allow_connection_outside_localhost() -> Result {
+    let mut sim = Builder::new().build();
+
+    sim.host("server", || async {
+        let listener = TcpListener::bind((Ipv4Addr::LOCALHOST, PORT)).await?;
+
+        loop {
+            let _ = listener.accept().await;
+        }
+    });
+
+    sim.client("client", async {
+        assert_error_kind(
+            TcpStream::connect(("server", PORT)).await,
+            io::ErrorKind::ConnectionRefused,
+        );
+
+        assert_error_kind(
+            TcpStream::connect((Ipv4Addr::LOCALHOST, PORT)).await,
+            io::ErrorKind::ConnectionRefused,
+        );
+
+        Ok(())
+    });
+
+    sim.run()
 }
```

Comment:
I would add another test that does at least one send and recv in each direction to test that logic. The tcp connection sequence is only in one direction since we cheat the syn-ack using a oneshot channel.